# Leveraging Twitter Data to Conquer the Digital Market of an Emerging Country (Togo)

## 1- Introduction


In this project, acting as a data analyst working for a digital marketing agency aiming to conquer the Togolese digital landscape, I have mined nearly 3,000 tweets from 1st January 2022 to 30 April 2023 containing the most popular hashtags among Togolese "Twittos" using the Python library Snscrape and also performed some analysis with SQL and finally creating vizualisations in Power BI to highlights insights from the analysis.


## 2- Data gathering from Twitter using Snscrape

In this section, after importing the necessary libraries, we have defined the most popular hashtags to analyze and the timeframe to be covered.
The scraped tweets are then stored in a list, converted in a dataframe and also saved as a csv file.

In [1]:
# Importing the libraries
import pandas as pd
import snscrape.modules.twitter as sntwitter
from datetime import datetime, timedelta

# Define the popular hashtags to analyze
hashtags = ['#Tgtwittos', '#TT228', '#Togo63', '#Team228', '#tgtwittos', '#team228']

# Define the start and end dates for the time frame
start_date = datetime(2022, 1, 1)
end_date = datetime(2023, 4, 30)

# Initialize an empty list to store the tweets
tweets_list = []

# Loop through each hashtag and collect the tweets
for hashtag in hashtags:
    query = f'{hashtag} lang:en'
    for tweet in sntwitter.TwitterSearchScraper(query + f' since:{start_date.date()} until:{end_date.date()}').get_items():
        tweets_list.append([tweet.id, tweet.user.id, hashtag, tweet.date, tweet.retweetCount, tweet.likeCount, tweet.replyCount, tweet.content, tweet.user.location,tweet.source])

# Convert the list of tweets into a Pandas dataframe
tweets_df = pd.DataFrame(tweets_list, columns=['tweet_id', 'user_id', 'hashtag', 'date', 'retweets', 'likes', 'replies', 'text', 'location','source'])

# Save the dataframe as a CSV file
tweets_df.to_csv('tweets.csv', index=False)

C:\Users\user\AppData\Local\Temp\ipykernel_9348\2332276814.py:20: DeprecatedFeatureWarning: content is deprecated, use rawContent instead
  tweets_list.append([tweet.id, tweet.user.id, hashtag, tweet.date, tweet.retweetCount, tweet.likeCount, tweet.replyCount, tweet.content, tweet.user.location,tweet.source])
Skipping unrecognised entry ID: 'promoted-tweet-1648959705727787008-2d49aba810ae77e'
Skipping unrecognised entry ID: 'promoted-tweet-1648959705727787008-2d44c4746843912'
Skipping unrecognised entry ID: 'promoted-tweet-1648959705727787008-2d4bb2c0994424b'
Skipping unrecognised entry ID: 'promoted-tweet-1648959705727787008-2d4d1b32bc9bb34'
Skipping unrecognised entry ID: 'promoted-tweet-1655949886884569091-2d4d48b816db8e3'
Skipping unrecognised entry ID: 'promoted-tweet-1648959705727787008-2d423c1a3c56961'
Skipping unrecognised entry ID: 'promoted-tweet-1655944095335067648-2d452c1cfc969b7'
Skipping unrecognised entry ID: 'promoted-tweet-1655944095335067648-2d4b31738f2da8f'
Skipping 

Skipping unrecognised entry ID: 'promoted-tweet-1648959705727787008-2d53c67f98e4785'
Unavailable user in card on tweet 1651615504010706951
Unavailable user in card on tweet 1651615504010706951
User 1541906272940830720 not found in user refs in card on tweet 1651615504010706951
User 1541906272940830720 not found in user refs in card on tweet 1651615504010706951
Skipping unrecognised entry ID: 'promoted-tweet-1648959705727787008-2d5809e20ec5970'
Skipping unrecognised entry ID: 'promoted-tweet-1651243541346107394-2d53351a9d19b91'
Stopping after 20 empty pages
Skipping unrecognised entry ID: 'promoted-tweet-1622883010977701888-2d5b64992fcee8f'
Skipping unrecognised entry ID: 'promoted-tweet-1648959705727787008-2d51888968e4e9e'
Skipping unrecognised entry ID: 'promoted-tweet-1648959705727787008-2d517c522657d8e'
Skipping unrecognised entry ID: 'promoted-tweet-1648959705727787008-2d5fc6c6b9f7030'
Skipping unrecognised entry ID: 'promoted-tweet-1655944095335067648-2d55d1f2bf3376f'
Skipping unr

In [2]:
tweets_df.head()

tweet_id              user_id     hashtag  \
0  1652407000083341314   940894696947306496  #Tgtwittos   
1  1652322746259456000  1078599142514003968  #Tgtwittos   
2  1652101793210421252           1902319314  #Tgtwittos   
3  1652081148854452225            555754465  #Tgtwittos   
4  1651981956022693889  1419331325274767363  #Tgtwittos   

                       date  retweets  likes  replies  \
0 2023-04-29 20:18:23+00:00         1      5        0   
1 2023-04-29 14:43:36+00:00         1      2        0   
2 2023-04-29 00:05:36+00:00         0      2        0   
3 2023-04-28 22:43:34+00:00         8     51        0   
4 2023-04-28 16:09:25+00:00         3      3        0   

                                                text                location  \
0  we are live link to teams : https://t.co/iCtDW...  Tech Enthusiast Palace   
1  " #Togo is a land of young innovators, doers, ...    République Togolaise   
2  The flag 🇹🇬                               The ...    République Togolaise   
3  The Flag                    /                 ...                    Togo   
4  HP PROBOOK 455 G9 NoteBook \nWindows 11Pro \nA...    République Togolaise   

  source  
0   None  
1   None  
2   None  
3   None  
4   None

## 3- Data cleaning and preprocessing

In [3]:
# first, let us check the shape of the dataset
tweets_df.shape

(2934, 10)

Our scraped tweets dataset covering the period contain 2934 rows and 10 columns.

In [4]:
# let's now check the data types of each column
tweets_df.dtypes

tweet_id                  int64
user_id                   int64
hashtag                  object
date        datetime64[ns, UTC]
retweets                  int64
likes                     int64
replies                   int64
text                     object
location                 object
source                   object
dtype: object

In [5]:
# let us count the number of null values in each column
null_counts = tweets_df.isnull().sum()

# print the null counts
print(null_counts)

tweet_id       0
user_id        0
hashtag        0
date           0
retweets       0
likes          0
replies        0
text           0
location       0
source      2934
dtype: int64


In [6]:
# check for duplicate rows
duplicates = tweets_df[tweets_df.duplicated()]

# print the duplicate rows
print(duplicates)

Empty DataFrame
Columns: [tweet_id, user_id, hashtag, date, retweets, likes, replies, text, location, source]
Index: []


It appears that there is no duplicate row in our dataset. This means there is no row with the exact same values in all the columns.

In [7]:
# let us check also if there are duplicate tweet_id
duplicated_id = tweets_df[tweets_df['tweet_id'].duplicated()]

# print duplicate tweet_id
print(duplicated_id)

                 tweet_id              user_id   hashtag  \
2290  1651600772822446082  1271383529050513408    #TT228   
2293  1647071038742855680  1559460640879955969    #TT228   
2298  1640037792670441476            273074763    #TT228   
2299  1638230044764938255            707143645    #TT228   
2300  1638229983863644166           2420813436    #TT228   
...                   ...                  ...       ...   
2925  1483106043274145797   702386763873042437  #Team228   
2926  1482730320730411011   702386763873042437  #Team228   
2927  1482445968574459908   702386763873042437  #Team228   
2929  1478654737708990468           1275885127  #Team228   
2931  1478322899597598726           1275885127  #Team228   

                          date  retweets  likes  replies  \
2290 2023-04-27 14:54:44+00:00         1      1        0   
2293 2023-04-15 02:55:11+00:00         0      2        0   
2298 2023-03-26 17:07:34+00:00         0      0        0   
2299 2023-03-21 17:24:14+00:00         


Our dataset has 190 rows with duplicate tweet_id, meaning there are 190 tweets with 2 or more hashtags.I choose not to drop the duplicated tweet_id rows as the analysis is based on the popularity of the hashtags there is no impartial way to drop a row of a specific hashtag in favour of another one without being biased.


In [8]:
# from the date column let us extratc the day of the week and the hour of the day and add them as news columns to the dataframe

# Convert the date column to a datetime format
tweets_df['date'] = pd.to_datetime(tweets_df['date'])

# Add a column for the day of the week and the hour of the day
tweets_df['day_of_week'] = tweets_df['date'].dt.day_name()
tweets_df['hour_of_day'] = tweets_df['date'].dt.hour

# we will use these two news columns further in our analysis of the engagement and popularity of tweets

In [9]:
tweets_df.head()

tweet_id              user_id     hashtag  \
0  1652407000083341314   940894696947306496  #Tgtwittos   
1  1652322746259456000  1078599142514003968  #Tgtwittos   
2  1652101793210421252           1902319314  #Tgtwittos   
3  1652081148854452225            555754465  #Tgtwittos   
4  1651981956022693889  1419331325274767363  #Tgtwittos   

                       date  retweets  likes  replies  \
0 2023-04-29 20:18:23+00:00         1      5        0   
1 2023-04-29 14:43:36+00:00         1      2        0   
2 2023-04-29 00:05:36+00:00         0      2        0   
3 2023-04-28 22:43:34+00:00         8     51        0   
4 2023-04-28 16:09:25+00:00         3      3        0   

                                                text                location  \
0  we are live link to teams : https://t.co/iCtDW...  Tech Enthusiast Palace   
1  " #Togo is a land of young innovators, doers, ...    République Togolaise   
2  The flag 🇹🇬                               The ...    République Togolaise   
3  The Flag                    /                 ...                    Togo   
4  HP PROBOOK 455 G9 NoteBook \nWindows 11Pro \nA...    République Togolaise   

  source day_of_week  hour_of_day  
0   None    Saturday           20  
1   None    Saturday           14  
2   None    Saturday            0  
3   None      Friday           22  
4   None      Friday           16

Let's look at the location column to see how it can be cleaned

Assuming most of the tweets come from users living in Togo,and for all the tweets where the location appears, we can clean the location column by searching for key words in a list and replacing them by the appropriate country name for the inconsistent values, let us define lists for each possible country to clean the column values.

In [42]:
# we will start by defining lists for each country and  iterate through them to check if the location value contains 
# any of the words in each list
def clean_location(location):
    togo_words = ['Togo', 'Palace','Tg', '', 'Republique togolaise', 'Lomé', 'lomé', 'TOGO', 
                  'LOME','Palace','127.0.0.2','🌎','Adétikopé',
                 '🇧🇯🇹🇬','Afrique','🇹🇬🇹🇬🇹🇬','Wam','Village','Anywhere','N/A','Aaa','Blackleonarddevinci',
                  'Heureux😊','.','','60','TK',
                 'osiris','Miller','Styx','Fatima','Metaverse','earth', 'part', 'Land','creative_land',
                  'African','m’aime','Africa',
                 'Somewhere','papa','multivers','papa','TOP','Rit', 'living','be','AFRICA','cieux🤴🏽',
                  'immatures', 'https://maps.app.goo.gl/NzFsRR',
                 'world','Minamadou','127.0.0.1', 'pussy','Fatima', 'Undefined', 'fan','Mind', 'nuage',
                  'mountain','you🏞','@hbk_trae','S.']
    benin_words = ['Bénin','Pobé','Hèvié','Natitingou']
    ghana_words = ['Kwadaso', 'Africa|WestAfrica|Ghana','ACCRA-GHANA']
    senegal_words = ['Sénégal','Dakar','🇸🇳🇹🇬🇹🇳️🇳🇬']
    usa_words = ['NY','IL','CA','WA','FL','TX','Kirkland,WA','CO','Indiana','DC','chicago🇬🇧',
                 'Floor.','unis', 'States','Monica']
    canada_words = ['Québec','Quebec','AB']
    england_words = ['London','London,England','Bridge']
    niger_words = ['Niger)']
    nigeria_words = ['🇳🇬']
    ci_words = ["d'Ivoire",'Abidjan','d’Ivoire','Bretagne' ]
    france_words = ['Lyon','Paris','Nancy']
    china_words = ['Shanghai','china']
    ben_words = ['Benelux']
    faso_words = ['Faso']
    kenya_words = ['Nairobi']
    ph_words = ['Peninsula']
    ma_words = ['Rabat']
    tk_words = ['Republic']
    jm_words = ['Jamaïque']
    gn_words = ['Guinée']
    words = location.split()
    for word in words:
        if word in togo_words:
            return 'Togo'
        elif word in benin_words:
            return 'Benin'
        elif word in ghana_words:
            return 'Ghana'
        elif word in senegal_words:
            return 'Senegal'
        elif word in usa_words:
            return 'USA'
        elif word in canada_words:
            return 'Canada'
        elif word in england_words:
            return 'England'
        elif word in niger_words:
            return 'Niger'
        elif word in nigeria_words:
            return 'Nigeria'
        elif word in ci_words:
            return 'Ivory Coast'
        elif word in france_words:
            return 'France'
        elif word in china_words:
            return 'China'
        elif word in ben_words:
            return 'Luxembourg'
        elif word in faso_words:
            return 'Burkina Faso'
        elif word in kenya_words:
            return 'Kenya'
        elif word in ph_words:
            return 'Phillipines'
        elif word in ma_words:
            return 'Maroc'
        elif word in tk_words:
            return 'Turkey'
        elif word in jm_words:
            return 'Jamaica'
        elif word in gn_words:
            return 'Guinea'
    return location

tweets_df['location'] = tweets_df['location'].apply(clean_location)
tweets_df.location.unique()

array(['Togo', '', 'USA', 'France', 'Ghana', 'Nigeria', 'Turkey',
       'Jamaica', 'China', 'Niger', 'Benin', 'Canada', 'Senegal',
       'Burkina Faso', 'Rwanda', 'Ivory Coast', 'India', 'Maroc',
       'Ethiopia', 'Guinea', 'Luxembourg', 'England', 'Mauritania',
       'Mali', 'Belgium', 'Germany', 'Pakistan', 'Kenya', 'Gabon',
       'Phillipines'], dtype=object)

In [44]:
# The source column will have been of a great value to gain insights on the most used platform but since it does not contain 
# any value, let us drop it.
tweets_df = tweets_df.drop('source', axis=1)
tweets_df.head()

tweet_id              user_id     hashtag  \
0  1652407000083341314   940894696947306496  #Tgtwittos   
1  1652322746259456000  1078599142514003968  #Tgtwittos   
2  1652101793210421252           1902319314  #Tgtwittos   
3  1652081148854452225            555754465  #Tgtwittos   
4  1651981956022693889  1419331325274767363  #Tgtwittos   

                       date  retweets  likes  replies  \
0 2023-04-29 20:18:23+00:00         1      5        0   
1 2023-04-29 14:43:36+00:00         1      2        0   
2 2023-04-29 00:05:36+00:00         0      2        0   
3 2023-04-28 22:43:34+00:00         8     51        0   
4 2023-04-28 16:09:25+00:00         3      3        0   

                                                text location day_of_week  \
0  we are live link to teams : https://t.co/iCtDW...     Togo    Saturday   
1  " #Togo is a land of young innovators, doers, ...     Togo    Saturday   
2  The flag 🇹🇬                               The ...     Togo    Saturday   
3  The Flag                    /                 ...     Togo      Friday   
4  HP PROBOOK 455 G9 NoteBook \nWindows 11Pro \nA...     Togo      Friday   

   hour_of_day  
0           20  
1           14  
2            0  
3           22  
4           16

In [46]:
# let us also drop the text column
tweets_df = tweets_df.drop('text', axis=1)

In [47]:
# We will end our data cleaning here by saving the dataframe in csv file
tweets_df.to_csv('tweets.csv', index=False)